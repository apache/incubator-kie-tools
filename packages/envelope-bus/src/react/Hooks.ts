/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiDefinition, MessageBusClient, NotificationPropertyNames, SubscriptionCallback } from "../api";
import { useEffect } from "react";

export function useSubscription<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    const subscription = bus.subscribe(method, callback);
    return () => {
      bus.unsubscribe(method, subscription);
    };
  }, [bus, method, callback]);
}

export function useSubscriptionOnce<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    let unsubscribed = false;

    const subscription = bus.subscribe(method, (...args) => {
      callback(...args);
      unsubscribed = true;
      bus.unsubscribe(method, subscription);
    });

    return () => {
      if (!unsubscribed) {
        bus.unsubscribe(method, subscription);
      }
    };
  }, [bus, method, callback]);
}
