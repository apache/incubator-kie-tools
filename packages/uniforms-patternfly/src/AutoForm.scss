/*
  This is required since there's a div between the form
  element and its children that breaks the following styles
    - display: grid
    - gap: var(--pf-c-form--GridGap)
  from @patternfly and making the fields very close to
  each other
*/
form[data-testid="autoform"] > div {
  display: inherit;
  gap: inherit;
}
