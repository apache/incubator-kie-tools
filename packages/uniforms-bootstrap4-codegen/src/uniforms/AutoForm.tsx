/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import escape from "lodash/escape";
import { Bridge } from "uniforms";
import { CodeGenElement } from "../api";
import { FORM, renderCodeGenElement } from "./templates/templates";
import { renderFormInputs } from "./rendering/RenderingUtils";

export type AutoFormProps = {
  id: string;
  disabled?: boolean;
  placeholder?: boolean;
  schema: Bridge;
};

const AutoForm: React.FC<AutoFormProps> = (props) => {
  const properties = {
    children: renderFormInputs(props.schema),
  };

  const form: CodeGenElement = renderCodeGenElement(FORM, properties);

  return <>{escape(form.html)}</>;
};

export default AutoForm;
