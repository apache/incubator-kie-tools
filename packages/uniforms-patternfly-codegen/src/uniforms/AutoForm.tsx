/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { Bridge } from "uniforms";
import union from "lodash/union";
import escape from "lodash/escape";
import * as prettier from "prettier";
import { FormElement } from "../api";
import { NS_SEPARATOR } from "./utils/Utils";
import { renderFormInputs } from "./rendering/RenderingUtils";
import { getStaticCodeBlock } from "./staticCode/staticCodeBlocks";

export type AutoFormProps = {
  id: string;
  disabled?: boolean;
  placeholder?: boolean;
  schema: Bridge;
};

const AutoForm: React.FC<AutoFormProps> = (props) => {
  const inputs: FormElement<any>[] = renderFormInputs(props.schema);

  let pfImports: string[] = [];
  let reactImports: string[] = [];
  let staticCodeArray: string[] = [];

  inputs.forEach((input) => {
    pfImports = union(pfImports, input.pfImports);
    reactImports = union(reactImports, input.reactImports);
    staticCodeArray = union(staticCodeArray, input.requiredCode);
  });

  const formName = `Form${props.id ? `${NS_SEPARATOR}${props.id}` : ""}`;
  const hooks = inputs.map((input) => input.stateCode).join("\n");
  const elements = inputs.map((input) => input.jsxCode).join("\n");
  const staticCodeStr: string = staticCodeArray.map((id) => getStaticCodeBlock(id)).join("\n");

  const formTemplate = `import React, { ${reactImports.join(", ")} }  from "react";
    import { ${pfImports.join(", ")} } from "@patternfly/react-core";
    
    const ${formName}: React.FC<any> = ( props:any ) => {
      ${hooks}
      
      ${staticCodeStr}
      
      return (
        <div className={'pf-c-form'}>
          ${elements}        
        </div>
      )    
    }
    
    export default ${formName};`;

  const formattedFormTemplate = prettier.format(formTemplate, {
    parser: "typescript",
    singleQuote: true,
    jsxSingleQuote: true,
    jsxBracketSameLine: true,
    tabWidth: 2,
  });
  return <>{escape(formattedFormTemplate)}</>;
};

export default AutoForm;
