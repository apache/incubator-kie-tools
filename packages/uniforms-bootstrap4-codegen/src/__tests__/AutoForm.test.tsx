/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { render } from "@testing-library/react";
import { unescape } from "lodash";
import AutoForm, { AutoFormProps } from "../uniforms/AutoForm";
import SimpleSchema from "simpl-schema";
import createSchema from "./_createSchema";

const schema = {
  personalData: { type: Object },
  "personalData.name": { type: String },
  "personalData.email": { type: String },
  address: { type: Object },
  "address.street": { type: String },
  "address.num": { type: SimpleSchema.Integer },
  "address.cp": { type: String },
  "address.city": { type: String },
  interview: { type: Object },
  "interview.position": {
    type: String,
    allowedValues: ["Developer", "HR", "UX"],
  },
  "interview.otherPositions": {
    type: Array,
    allowedValues: ["Developer", "HR", "UX"],
  },
  "interview.otherPositions.$": String,
  "interview.skills": {
    type: Array,
    allowedValues: ["Java", "React", "TypeScript", "Quarkus"],
    uniforms: { checkboxes: true },
  },
  "interview.skills.$": String,
  "interview.age": {
    type: SimpleSchema.Integer,
    min: 18,
    max: 99,
  },
  "interview.salary": {
    type: Number,
    min: 0,
    max: 1000.5,
  },
  "interview.rating": {
    type: Number,
    allowedValues: [1, 2, 3, 4, 5],
    uniforms: { checkboxes: true },
  },
  "interview.hire": Boolean,
  "interview.hidingDate": Date,
  friends: { type: Array },
  "friends.$": Object,
  "friends.$.name": { type: String },
  "friends.$.age": { type: Number },
};

const props: AutoFormProps = {
  id: "HRInterview",
  schema: createSchema(schema),
  disabled: false,
  placeholder: true,
};

describe("<AutoForm> tests", () => {
  it("<AutoForm> - Full rendering", () => {
    const { container } = render(<AutoForm {...props} />);

    expect(container).toMatchSnapshot();
    const formSource = unescape(container.innerHTML);
    expect(formSource).not.toBeUndefined();
  });
});
