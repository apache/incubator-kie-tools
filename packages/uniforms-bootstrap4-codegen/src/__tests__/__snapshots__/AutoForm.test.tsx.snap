// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoForm> tests <AutoForm> - Full rendering 1`] = `
<div>
  &lt;div&gt;
  &lt;fieldset&gt;
    &lt;legend&gt;Personal data&lt;/legend&gt;
    &lt;div&gt;
      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-0002&quot;&gt;Name&lt;/label&gt;
        &lt;input type=&quot;text&quot; id=&quot;uniforms-0000-0002&quot; name=&quot;personalData.name&quot; class=&quot;form-control&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-0003&quot;&gt;Email&lt;/label&gt;
        &lt;input type=&quot;text&quot; id=&quot;uniforms-0000-0003&quot; name=&quot;personalData.email&quot; class=&quot;form-control&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;
    &lt;/div&gt;
  &lt;/fieldset&gt;

  &lt;fieldset&gt;
    &lt;legend&gt;Address&lt;/legend&gt;
    &lt;div&gt;
      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-0006&quot;&gt;Street&lt;/label&gt;
        &lt;input type=&quot;text&quot; id=&quot;uniforms-0000-0006&quot; name=&quot;address.street&quot; class=&quot;form-control&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-0008&quot;&gt;Num&lt;/label&gt;
        &lt;input type=&quot;number&quot; class=&quot;form-control&quot; id=&quot;uniforms-0000-0008&quot; name=&quot;address.num&quot; step=&quot;1&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-0009&quot;&gt;Cp&lt;/label&gt;
        &lt;input type=&quot;text&quot; id=&quot;uniforms-0000-0009&quot; name=&quot;address.cp&quot; class=&quot;form-control&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000a&quot;&gt;City&lt;/label&gt;
        &lt;input type=&quot;text&quot; id=&quot;uniforms-0000-000a&quot; name=&quot;address.city&quot; class=&quot;form-control&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;
    &lt;/div&gt;
  &lt;/fieldset&gt;

  &lt;fieldset&gt;
    &lt;legend&gt;Interview&lt;/legend&gt;
    &lt;div&gt;
      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000e&quot;&gt;Position&lt;/label&gt;
        &lt;select class=&quot;form-control&quot; id=&quot;uniforms-0000-000e&quot; name=&quot;interview.position&quot;&gt;
          &lt;option value=&quot;Developer&quot;&gt;Developer&lt;/option&gt;

          &lt;option value=&quot;HR&quot;&gt;HR&lt;/option&gt;

          &lt;option value=&quot;UX&quot;&gt;UX&lt;/option&gt;
        &lt;/select&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000g&quot;&gt;Other positions&lt;/label&gt;
        &lt;select class=&quot;form-control&quot; id=&quot;uniforms-0000-000g&quot; name=&quot;interview.otherPositions&quot; multiple&gt;
          &lt;option value=&quot;Developer&quot;&gt;Developer&lt;/option&gt;

          &lt;option value=&quot;HR&quot;&gt;HR&lt;/option&gt;

          &lt;option value=&quot;UX&quot;&gt;UX&lt;/option&gt;
        &lt;/select&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000i&quot;&gt;Skills&lt;/label&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;checkbox&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000i__Java&quot; name=&quot;interview.skills&quot; value=&quot;Java&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000i__Java&quot;&gt;Java&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;checkbox&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000i__React&quot; name=&quot;interview.skills&quot; value=&quot;React&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000i__React&quot;&gt;React&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;checkbox&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000i__TypeScript&quot; name=&quot;interview.skills&quot; value=&quot;TypeScript&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000i__TypeScript&quot;&gt;TypeScript&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;checkbox&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000i__Quarkus&quot; name=&quot;interview.skills&quot; value=&quot;Quarkus&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000i__Quarkus&quot;&gt;Quarkus&lt;/label&gt;
        &lt;/div&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000k&quot;&gt;Age&lt;/label&gt;
        &lt;input type=&quot;number&quot; class=&quot;form-control&quot; id=&quot;uniforms-0000-000k&quot; name=&quot;interview.age&quot; min=&quot;18&quot; max=&quot;99&quot; step=&quot;1&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000m&quot;&gt;Salary&lt;/label&gt;
        &lt;input type=&quot;number&quot; class=&quot;form-control&quot; id=&quot;uniforms-0000-000m&quot; name=&quot;interview.salary&quot; max=&quot;1000.5&quot; step=&quot;0.01&quot; value=&quot;&quot; /&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000o&quot;&gt;Rating&lt;/label&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;radio&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000o__1&quot; name=&quot;interview.rating&quot; value=&quot;1&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000o__1&quot;&gt;1&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;radio&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000o__2&quot; name=&quot;interview.rating&quot; value=&quot;2&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000o__2&quot;&gt;2&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;radio&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000o__3&quot; name=&quot;interview.rating&quot; value=&quot;3&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000o__3&quot;&gt;3&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;radio&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000o__4&quot; name=&quot;interview.rating&quot; value=&quot;4&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000o__4&quot;&gt;4&lt;/label&gt;
        &lt;/div&gt;

        &lt;div class=&quot;form-check&quot;&gt;
          &lt;input type=&quot;radio&quot; class=&quot;form-check-input&quot; id=&quot;uniforms-0000-000o__5&quot; name=&quot;interview.rating&quot; value=&quot;5&quot; /&gt;
          &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000o__5&quot;&gt;5&lt;/label&gt;
        &lt;/div&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-check&quot;&gt;
        &lt;input type=&quot;checkbox&quot; id=&quot;uniforms-0000-000q&quot; name=&quot;interview.hire&quot; class=&quot;form-check-input&quot; /&gt;
        &lt;label class=&quot;form-check-label&quot; for=&quot;uniforms-0000-000q&quot;&gt;Hire&lt;/label&gt;
      &lt;/div&gt;

      &lt;div class=&quot;form-group&quot;&gt;
        &lt;label for=&quot;uniforms-0000-000s&quot;&gt;Hiding date&lt;/label&gt;
        &lt;input type=&quot;datetime-local&quot; id=&quot;uniforms-0000-000s&quot; name=&quot;interview.hidingDate&quot; class=&quot;form-control&quot; /&gt;
      &lt;/div&gt;
    &lt;/div&gt;
  &lt;/fieldset&gt;

  &lt;div class=&quot;alert alert-warning&quot; role=&quot;alert&quot;&gt;
    &lt;h4 class=&quot;alert-heading&quot;&gt;Unsupported field type: Array&lt;/h4&gt;
    &lt;p&gt;Cannot find form control for property &lt;code&gt;friends&lt;/code&gt; with type &lt;code&gt;Array&lt;/code&gt;.&lt;/p&gt;
    &lt;hr /&gt;
    &lt;p class=&quot;mb-0&quot;&gt;
      Some complex property types, such as &lt;code&gt;Array&lt;/code&gt; aren&#39;t yet supported, however, you can still write your own component into the
      form and manually bind it.
    &lt;/p&gt;
  &lt;/div&gt;
&lt;/div&gt;

</div>
`;
